internal/core/future.rs

```diff
@@ -0,0 +1,168 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.0 OR LicenseRef-Slint-commercial
+
+#![cfg(target_has_atomic = "ptr")] // Arc is not available. TODO: implement using RawWarker
+#![warn(missing_docs)]
+
+//! This module contains the code that runs futures
+
+use crate::api::{invoke_from_event_loop, EventLoopError};
+use alloc::boxed::Box;
+use alloc::task::Wake;
+use alloc::vec::Vec;
+use atomic_polyfill as atomic;
+use core::future::Future;
+use core::ops::DerefMut;
+use core::pin::Pin;
+use core::task::Poll;
+
+enum FutureState<T> {
+    Running(Pin<Box<dyn Future<Output = T>>>),
+    Finished(Option<T>),
+}
+
+struct FutureRunnerInner<T> {
+    fut: FutureState<T>,
+    wakers: Vec<core::task::Waker>,
+}
+
+struct FutureRunner<T> {
+    #[cfg(not(feature = "std"))]
+    inner: core::cell::RefCell<FutureRunnerInner<T>>,
+    #[cfg(feature = "std")]
+    inner: std::sync::Mutex<FutureRunnerInner<T>>,
+    aborted: atomic::AtomicBool,
+
+    #[cfg(feature = "std")]
+    thread: std::thread::ThreadId,
+}
+
+impl<T> FutureRunner<T> {
+    fn inner(&self) -> impl DerefMut<Target = FutureRunnerInner<T>> + '_ {
+        #[cfg(feature = "std")]
+        return self.inner.lock().unwrap();
+        #[cfg(not(feature = "std"))]
+        return self.inner.borrow_mut();
+    }
+}
+
+// # Safety:
+// The Future might not be Send, but we only poll the future from the main thread.
+// (We even assert that)
+// We may access the finished value from another thread only if T is Send
+// (because JoinHandle only implement Send if T:Send)
+#[allow(unsafe_code)]
+unsafe impl<T> Send for FutureRunner<T> {}
+#[allow(unsafe_code)]
+unsafe impl<T> Sync for FutureRunner<T> {}
+
+impl<T: 'static> Wake for FutureRunner<T> {
+    fn wake(self: alloc::sync::Arc<Self>) {
+        invoke_from_event_loop(move || {
+            #[cfg(feature = "std")]
+            assert_eq!(self.thread, std::thread::current().id(), "the future was moved to a thread despite we checked it was created in the event loop thread");
+            let waker = self.clone().into();
+            let mut inner = self.inner();
+            let mut cx = core::task::Context::from_waker(&waker);
+            if let FutureState::Running(fut) = &mut inner.fut {
+                if self.aborted.load(atomic::Ordering::Relaxed) {
+                    inner.fut = FutureState::Finished(None);
```

Comment:
that's why i was thinking we should make JoinHandle::abort() take `self` by value, so it can't be polled again, by construction.  (and it isn't clone either right now)

Reply:
That's a compelling argument. Still feel like changing it? :)